File: Pset-1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main (void)
{
    //prompt the user for input
    long n;
    do 
    {
        n = get_long ("Number: ");
    }   while ((n < 12) && (n > 16));

    // check the length and the first digits
    int len, f, fd; //len = length; f = first digit; fd = 2 first digits
    len = (int)log10(n);
    f = (int)(n / pow(10, len));
    fd = (int)(n / pow(10, len - 1));
    
    //checksum
    int s1, r1; // s = sum; r = remainder
    s1 = 0;
    int r2, i, r3, s2 = 0, s3 = 0; // r2 = remainder; i = value of the digits multiplied by 2; r3 = remainder of the digits like 5,6,7 when *2 have to digits; s2 = sum 
                                 // s3 = sum of the digits like 5,6,7
    while (n != 0)
    {
        r1 = n % 10;
        s1 = s1 + r1;
        r2 = (n/10) % 10;
        i = r2 * 2;
        
            if (i >= 10)
                while (i != 0)
                {
                    r3 = i % 10;
                    s3 = s3 + r3;
                    i = i / 10;
                }
            else
                s3 = i;

        s2 = s2 + s3;
        n = n / 100;
    }
    
    int j; //j = the last digit of the sum
    j = ((s1 + s2) % 10);
    
    if ((f == 4) && (j == 0))
        printf ("VISA\n");
    else if ((f != 4) && (fd == 34 || fd == 37) && (j == 0))
        printf ("AMEX\n");
    else if ((f != 4) && (fd == 51 || fd == 52 || fd == 53 || fd == 54 || fd ==55) && (j == 0))
        printf ("MASTERCARD\n");
    else
        printf ("INVALID\n");
}

File: Pset-6/credit.py
from math import pow

def main():
    #get the input
    card_number = get_long()

    if check_valid(card_number):
        print_brand(card_number)
    else:
        print("INVALID")
    
def get_long():
    while True:
        try:
            number = int(input("Number: "))
            if number > 0:
                break
        except Exception as e:
            print(e)
    return number

def check_valid(n):
    return check_sum(n)   

def check_sum(n):
    sum = 0
    num_digits = len(str(n))
    #if the num of digits is an odd number(13 or 15) 
    #we will need to *2 the even digits and add sum with the odd digits
    if num_digits == 16:
        for i in range(num_digits):
            if (i % 2 == 0):
                digit = 2 * (n % 10)
                sum += digit // 10 + digit % 10
            else:
                sum += n % 10
            n //= 10

    #for the even number (16) digits, we *2 the odd digits and add sum the even digits            
    elif num_digits == 13 or num_digits == 15:
        for i in range(num_digits):
            if (i % 2 == 0):
                sum += n % 10
            else:
                digit = 2 * (n % 10)
                sum += digit // 10 + digit % 10
            n //= 10
    return sum % 10 == 0

def print_brand(n):
    if (n >= 34e13 and n < 35e13) or (n >= 37e13 and n < 38e13):
        print("AMEX")
    elif n >= 51e14 and n <= 56e14:
        print("MASTERCARD")
    elif (n >= 4e12 and n < 5e12) or (n >= 4e15 and n < 5e15):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
